// uu5g04/dist-node/bricks/todo.js
import React from 'react';
import BaseMixin from '../common/base-mixin.js';
import ElementaryMixin from '../common/elementary-mixin.js';
import SectionMixin from '../common/section-mixin.js';
import NestingLevelMixin from '../common/nesting-level-mixin.js';
import Tools from '../common/tools.js';

import './todo.less';

export const Todo = React.createClass({

  //@@viewOn:mixins
  mixins: [
    BaseMixin,
    ElementaryMixin,
    SectionMixin,
    NestingLevelMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'UU5.Bricks.Todo',
    nestingLevel: 'box',
    classNames: {
      main: 'uu5-bricks-todo'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    props: React.PropTypes.oneOfType([
      React.PropTypes.array,
      React.PropTypes.object
    ]),
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      props: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return (
      this.getNestingLevel()
        ? (
        <UU5.Bricks.Section {...this.getMainPropsToPass()} content={null}>
          <UU5.Bricks.Pre>
            {this.props.props && Tools.prettyJson(this.props.props)}
          </UU5.Bricks.Pre>
        </UU5.Bricks.Section>
      ) : null
    );
  }
  //@@viewOff:render
});

export default Todo;
